Add `better-boolean-variable-names` (#2446)
sindresorhus/eslint-plugin-unicorn: docs/rules/better-boolean-variable-names.md

> # Prefer readable Boolean variable names

Reviewer: ```suggestion
# Prefer readable boolean variable names
```

💼 This rule is enabled in the ✅ `recommended` [config](https://github.com/sindresorhus/eslint-plugin-unicorn#preset-configs-eslintconfigjs).

💡 This rule is manually fixable by [editor suggestions](https://eslint.org/docs/latest/use/core-concepts#rule-suggestions).

<!-- end auto-generated rule header -->
<!-- Do not manually modify this header. Run: `npm run fix:eslint-docs` -->

> When it is clear that an expression is a Boolean value, the variable name should start with `is`/`was`/`has`/`can`/`should` to improve readability.

Reviewer 2: `will` also.

> When it is clear that an expression is a Boolean value, the variable name should start with `is`/`was`/`has`/`can`/`should` to improve readability.

Reviewer: ```suggestion
Boolean variable names should start with `is`/`was`/`has`/`can`/`should` for clarity and readability.
```

## Examples

```js
const completed = true; // ❌
const isCompleted = true; // ✅
```

```js
const completed = progress === 100; // ❌
const isCompleted = progress === 100; // ✅
```

```js
const completed = Boolean('true'); // ❌
const isCompleted = Boolean('true'); // ✅
```

```js
const completed = new Boolean('true'); // ❌
const isCompleted = new Boolean('true'); // ✅
```

```js
const adult = age >= 18; // ❌
const isAdult = age >= 18; // ✅
```

```js
const adult = age >= 18 ? true : false; // ❌
const isAdult = age >= 18 ? true : false; // ✅
```

```js
const gotModifyRights = isGotPreviewRights() && isGotDownloadRights(); // ❌

> const isGotModifyRights = isGotPreviewRights() && isGotDownloadRights(); // ✅

Reviewer 2: Why? I think got is right. 

``` js
const gotMilk = true
```

Author: **Here is the response from AI:**

The word "got" is generally not suitable as a prefix for Boolean variables because it does not conform to the conventions for naming Boolean variables. Boolean variable names typically use verbs or adverbs to clearly express their Boolean value, such as "is," "has," "can," etc.

Using "got" as a prefix may result in variable names that are not sufficiently clear, thereby reducing the readability and maintainability of the code. Therefore, it is recommended to avoid using "got" as a prefix for Boolean variables.

Better choices include the following prefixes:

- `is`: isUserLoggedIn
- `has`: hasAdminPrivileges
- `can`: canEditProfile
- `should`: shouldDisplayNotification
- `will`: willSendEmail
- `needs`: needsPasswordReset

These prefixes can more clearly express the meaning of Boolean variables.

const downloaded = isCompleted

// ✅
const isCompleted = true
const isDownloaded = isCompleted
```

<!-- Type Annotation -->
## Type Annotation

```js
const completed = isCompleted as  boolean; // ❌
const isCompleted = isCompleted as  boolean; // ✅
```

```js
const completed = isCompleted() as  boolean; // ❌
const isCompleted = isCompleted() as  boolean; // ✅
```

```js
// ❌
var isCompleted: boolean
const downloaded = isCompleted

// ✅
var isCompleted: boolean
const isDownloaded = isCompleted
```

```js
// ❌
function isCompleted(): boolean {}
const downloaded = isCompleted()

// ✅
function isCompleted(): boolean {}
const isDownloaded = isCompleted()
```

```js
function completed(): boolean {} // ❌
function isCompleted(): boolean {} // ✅
```

```js
const completed = (): boolean => {} // ❌
const isCompleted = (): boolean => {} // ✅
```

```js
function download(url: string, showProgress: boolean) {} // ❌

function download(url: string, shouldShowProgress: boolean) {} // ✅
```

## Options

You can specify which words can start with the string. The default is `['is', 'was', 'has', 'can', 'should']`.

If specified, it will be merged with the default value.

```js
{

>   'unicorn/prefer-readable-boolean-variable-names': [

Reviewer: Tab-indentation.

    'error',
    {

>       prefixes: ['will', 'allows']

Reviewer 2: This gives me doubts about this rule. Having to configure your own verbs is a smell to me. `allows` should be valid by default… maybe? Or should it be `doesAllowDeletion`? What situations are we missing?

I don’t really want a rule that bugs you for names that are fine and that might be difficult to get right.

Author: How about we add more prefixes to covering most scenarios? eg.

```js
["is","was","has","had","will","would","can","could","should","shall","does","needs","must","includes","enables","supports"]
```
